### PR TITLE
Implementation of PAM encoder/decoder

### PR DESCRIPTION
**Pull request by DmtKats**
This pull request implements the PAM image format as defined at:
http://netpbm.sourceforge.net/doc/pam.html

The PAM format provides a generic means for storing 2 dimensional information.
This is useful for opencv since there are cases where data gets translated into
non standardized formats, which makes it difficult to store and load this information.

### CODE REVIEW

File: modules/imgcodecs/src/grfmt_pam.cpp [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+
+    int width = img.cols, height = img.rows;
+    int stride = width*(int)img.elemSize();
+    const uchar* data = img.ptr();
+    const struct pam_format *fmt = NULL;
+    int x, y, tmp, bufsize = 256;
+
+    /* parse save file type */
+    for( size_t i = 0; i < params.size(); i += 2 )
+        if( params[i] == CV_IMWRITE_PXM_BINARY ) {
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_pam.cpp]: `IMWRITE_PAM_TUPLETYPE`?

**Comment by vpisarev** [comment]
@DmtKats, thanks for the contribution! It looks like some build slaves fail. Please, fix it

**Comment by ribalda** [comment]
The build errors seems to be time timeout while fetching/merging:

```
command timed out: 1200 seconds without output running ['git', 'fetch', '-t', 'git://code.ocv/opencv.git', 'master'], attempting to kill
program finished with exit code 1
elapsedTime=1200.939000

```

```
POST git-upload-pack (957 bytes)
fatal: The remote end hung up unexpectedly
error: RPC failed; result=52, HTTP code = 0
```

@vpisarev are you sure that it is not a server hiccup?

Thanks!

**Comment by alalek** [comment]
These builds were restarted, but there is merge conflict. Please rebase patch and resolve merge conflict (via `git pull --rebase`).

**Comment by DmtKats** [comment]
Hmm, the test error I got seems to not be related with the code I have added. Can someone verify that that test does not use anything from the imgcodecs module? I checked the source but didn't find any functions pertaining my changed but I might be wrong.

**Comment by alalek** [comment]
@DmtKats #6950 should fix this problem

**Comment by DmtKats** [comment]
Ok, so I should wait for the other pull request to be applied before triggering a new build?

**Comment by vpisarev** [comment]
@DmtKats, looks like now all the builders are "green". Before merging the PR, may I ask you to add some test for PAM files to opencv/modules/imgcodecs/test/test_grfmt.cpp? Basically, you can, for example, copy the jpeg test and modify it.

**Comment by DmtKats** [comment]
I have implemented a simple set of tests for PAM. A requirement for the tests to pass is that this pull request in extras gets accepted: https://github.com/opencv/opencv_extra/pull/304

It adds a PAM image of lena.

**Comment by alalek** [comment]
@DmtKats Please reopen both PRs with other branch name (for example, "pam"), because OpenCV test infrastructure ignores branches called "master" or "2.4".

**Pull request closed without merging**